executor/write.go

```diff
@@ -817,59 +834,60 @@ func (e *InsertExec) exec(ctx context.Context, rows [][]types.Datum) (Row, error
 	if len(e.OnDuplicate) > 0 && !e.IgnoreErr {
-		var err error
-		rows, err = e.batchUpdateDupRows(rows, e.OnDuplicate)
-		if err != nil {
-			return nil, errors.Trace(err)
-		}
-	} else if len(e.OnDuplicate) == 0 && e.IgnoreErr {
-		// If you use the IGNORE keyword, duplicate-key error that occurs while executing the INSERT statement are ignored.
-		// For example, without IGNORE, a row that duplicates an existing UNIQUE index or PRIMARY KEY value in
-		// the table causes a duplicate-key error and the statement is aborted. With IGNORE, the row is discarded and no error occurs.
-		// However, if the `on duplicate update` is also specified, the duplicated row will be updated.
-		// Using BatchGet in insert ignore to mark rows as duplicated before we add records to the table.
-		var err error
-		rows, err = batchMarkDupRows(e.ctx, e.Table, rows)
+		err := e.batchUpdateDupRows(rows, e.OnDuplicate)
 		if err != nil {
 			return nil, errors.Trace(err)
 		}
-	}
-	for _, row := range rows {
-		// duplicate row will be marked as nil in batchMarkDupRows if
-		// IgnoreErr is true. For IgnoreErr is false, it is a protection.
-		if row == nil {
-			continue
-		}
-		if err := e.checkBatchLimit(); err != nil {
-			return nil, errors.Trace(err)
-		}
-		if len(e.OnDuplicate) == 0 && !e.IgnoreErr {
-			e.ctx.Txn().SetOption(kv.PresumeKeyNotExists, nil)
-		}
-		h, err := e.Table.AddRecord(e.ctx, row, false)
-		e.ctx.Txn().DelOption(kv.PresumeKeyNotExists)
-		if err == nil {
-			if !sessVars.ImportingData {
-				e.ctx.StmtAddDirtyTableOP(DirtyTableAddRow, e.Table.Meta().ID, h, row)
+	} else {
+		if len(e.OnDuplicate) == 0 && e.IgnoreErr {
+			// If you use the IGNORE keyword, duplicate-key error that occurs while executing the INSERT statement are ignored.
+			// For example, without IGNORE, a row that duplicates an existing UNIQUE index or PRIMARY KEY value in
+			// the table causes a duplicate-key error and the statement is aborted. With IGNORE, the row is discarded and no error occurs.
+			// However, if the `on duplicate update` is also specified, the duplicated row will be updated.
+			// Using BatchGet in insert ignore to mark rows as duplicated before we add records to the table.
+			var err error
+			rows, err = batchMarkDupRows(e.ctx, e.Table, rows)
+			if err != nil {
+				return nil, errors.Trace(err)
 			}
-			e.rowCount++
-			continue
 		}
-		if kv.ErrKeyExists.Equal(err) {
-			// TODO: Use batch get to speed up `insert ignore on duplicate key update`.
-			if len(e.OnDuplicate) > 0 && e.IgnoreErr {
-				data, err1 := e.Table.RowWithCols(e.ctx, h, e.Table.WritableCols())
-				if err1 != nil {
-					return nil, errors.Trace(err1)
-				}
-				if _, _, _, err = e.doDupRowUpdate(h, data, row, e.OnDuplicate); err != nil {
-					return nil, errors.Trace(err)
+		for _, row := range rows {
+			// duplicate row will be marked as nil in batchMarkDupRows if
+			// IgnoreErr is true. For IgnoreErr is false, it is a protection.
```

Comment:
“For IgnoreErr being false, it is a protection." ? Please also specify what is this protection.

Reply:
`batchMarkDupRows` will set row = nil to mark the row is duplicated, so we need to skip the nil row here. If ignore error is false, the row should not be nil, but we still check  
 nil here, so it is the protection.